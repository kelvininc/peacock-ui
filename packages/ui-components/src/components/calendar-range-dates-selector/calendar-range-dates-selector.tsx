import { Component, Event, EventEmitter, h, Host, Prop, Watch } from '@stencil/core';
import { formatDate, fromISOToMoment, getDatesBetweenRange, isDateBefore, isDateSame, isDateValid } from '../../utils/date.helper';
import { IClickDateEvent, SelectedRange } from '../calendar/calendar.types';
import { ICalendarRangeDatesSelector, ICalendarRangeDatesSelectorEvents, ISelectRangeDates } from './calendar-range-dates-selector.types';

@Component({
	tag: 'kv-calendar-range-dates-selector',
	styleUrls: {
		night: 'calendar-range-dates-selector.night.scss',
		light: 'calendar-range-dates-selector.light.scss'
	},
	shadow: true
})
export class KvCalendarRangeDatesSelector implements ICalendarRangeDatesSelector, ICalendarRangeDatesSelectorEvents {
	/** @inheritdoc */
	@Prop({ reflect: false }) selectedRangeDates?: SelectedRange = [];
	/** @inheritdoc */
	@Prop({ reflect: false }) initialDate?: string;
	/** @inheritdoc */
	@Prop({ reflect: false }) disabledDates?: string[] = [];
	/** @inheritdoc */
	@Prop({ reflect: false }) minDate?: string;
	/** @inheritdoc */
	@Prop({ reflect: false }) maxDate?: string;

	/** @inheritdoc */
	@Event() selectRangeDates: EventEmitter<ISelectRangeDates>;

	@Watch('selectedRangeDates')
	validateRangeDates(newRangeDates: SelectedRange | undefined) {
		if (newRangeDates !== undefined && newRangeDates.length > 0) {
			const [selectedStartDate, selectedEndDate] = newRangeDates;

			if (selectedStartDate !== undefined && !isDateValid(selectedStartDate)) {
				throw new Error('Range selected start date should be a valid date');
			}

			if (selectedEndDate !== undefined && !isDateValid(selectedEndDate)) {
				throw new Error('Range selected end date should be a valid date');
			}
		}
	}

	componentWillLoad() {
		this.validateRangeDates(this.selectedRangeDates);
	}

	private onClickDate = ({ detail: { event, payload: date } }: CustomEvent<IClickDateEvent>): void => {
		const clickedDateMoment = fromISOToMoment(date);

		const [selectedStartDate, selectedEndDate] = this.getSelectedRangeDates();

		// check if start date is not selected
		if (selectedStartDate === undefined) {
			this.selectRangeDates.emit({ event, payload: [formatDate(clickedDateMoment)] });
			return;
		}

		// check if clicked date is the same as the selected date
		if (isDateSame(clickedDateMoment, selectedStartDate)) {
			// check if end date is not selected
			if (selectedEndDate === undefined) {
				this.selectRangeDates.emit({ event, payload: [formatDate(clickedDateMoment), formatDate(clickedDateMoment)] });
				return;
			}

			this.selectRangeDates.emit({ event, payload: [] });
			return;
		}

		// check if end date is selected
		if (selectedEndDate !== undefined) {
			// reset end date and set
			// start date to the clicked date
			this.selectRangeDates.emit({ event, payload: [formatDate(clickedDateMoment)] });
			return;
		}

		// check if clicked day is before the start date
		if (isDateBefore(clickedDateMoment, selectedStartDate)) {
			// reset end date and set
			// start date to clicked date
			this.selectRangeDates.emit({ event, payload: [formatDate(clickedDateMoment)] });
			return;
		}

		// the clicked date is after the clicked date
		// set end date to the clicked day
		this.selectRangeDates.emit({ event, payload: [formatDate(selectedStartDate), formatDate(clickedDateMoment)] });
		return;
	};

	public getSelectedDates = (): string[] | undefined => {
		const [selectedStartDate, selectedEndDate] = this.getSelectedRangeDates();

		if (selectedStartDate === undefined) {
			return;
		}

		if (selectedEndDate === undefined) {
			return [selectedStartDate];
		}

		return getDatesBetweenRange(selectedStartDate, selectedEndDate);
	};

	private getSelectedRangeDates = () => this.selectedRangeDates ?? [];

	render() {
		const [selectedStartDate] = this.getSelectedRangeDates();
		const selectedDates = this.getSelectedDates();

		return (
			<Host>
				<kv-calendar
					initialDate={this.initialDate ?? selectedStartDate}
					onClickDate={this.onClickDate}
					selectedDates={selectedDates}
					disabledDates={this.disabledDates}
					minDate={this.minDate}
					maxDate={this.maxDate}
					exportparts="calendar-container,month-container,day-container"
				/>
			</Host>
		);
	}
}
