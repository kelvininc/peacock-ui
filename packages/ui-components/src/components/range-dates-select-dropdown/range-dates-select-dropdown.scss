@import '../../assets/styles/globals';

.range-dates-select-dropdown {
	user-select: none;
}

.start-single-date-select-dropdown {
	margin-bottom: $spacing-5x;
}

.range-dates-select-dropdown-calendar-container {
	padding-top: $spacing-6x;
}
