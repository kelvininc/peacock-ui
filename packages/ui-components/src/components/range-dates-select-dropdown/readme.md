# kv-range-dates-select-dropdown

<!-- Auto Generated Below -->


## Usage

### Angular

```html
<!-- Default -->
<kv-range-dates-select-dropdown></kv-range-dates-select-dropdown>

<!-- With input configs -->
<kv-range-dates-select-dropdown
	[startInputConfig]="{ label: 'From', placeholder: 'Select a start date' }"
	[endInputConfig]="{ label: 'To', placeholder: 'Select an end date date' }"
></kv-range-dates-select-dropdown>
```


### React

```tsx
import React from 'react';

import { KvRangeDatesSelectDropdown } from '@kelvininc/react-ui-components';

export const KvRangeDatesSelectDropdownExample: React.FC = () => (
	<>
		{/*-- Default --*/}
		<KvRangeDatesSelectDropdown />

		{/*-- With input config --*/}
		<KvRangeDatesSelectDropdown
			startInputConfig={{ label: 'From', placeholder: 'Select a start date' }}
			endInputConfig={{ label: 'From', placeholder: 'Select an end date' }}
		/>
	</>
);
```



## Properties

| Property                  | Attribute      | Description                                                                                         | Type                                                                                              | Default                             |
| ------------------------- | -------------- | --------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------- | ----------------------------------- |
| `autoClose`               | `auto-close`   | (optional) If `true` the dropdown will be automatically closed after start and end date is selected | `boolean`                                                                                         | `true`                              |
| `disabledDates`           | --             | (options) The disabled dates array                                                                  | `string[]`                                                                                        | `[]`                                |
| `dropdownPositionOptions` | --             | (optional) The dropdown position options                                                            | `{ platform?: Platform; placement?: Placement; strategy?: Strategy; middleware?: Middleware[]; }` | `DEFAULT_DROPDOWN_POSITION_OPTIONS` |
| `endInputConfig`          | --             | (optional) The end date text field input custom configurations                                      | `IInputConfig`                                                                                    | `{}`                                |
| `initialDate`             | `initial-date` | (optional) Initial date                                                                             | `string`                                                                                          | `undefined`                         |
| `isOpen`                  | `is-open`      | (optional) If `true` the calendar is opened                                                         | `boolean`                                                                                         | `false`                             |
| `maxDate`                 | `max-date`     | (optional) Maximum accepted date                                                                    | `string`                                                                                          | `undefined`                         |
| `minDate`                 | `min-date`     | (optional) Minimum accepted date                                                                    | `string`                                                                                          | `undefined`                         |
| `selectedRangeDates`      | --             | (optional) Currently selected range dates                                                           | `[] \| [string, string] \| [string]`                                                              | `undefined`                         |
| `startInputConfig`        | --             | (optional) The start date text field input custom configurations                                    | `IInputConfig`                                                                                    | `{}`                                |


## Events

| Event              | Description                                            | Type                             |
| ------------------ | ------------------------------------------------------ | -------------------------------- |
| `openStateChange`  | Emitted when the calendar selector opens state changes | `CustomEvent<boolean>`           |
| `selectRangeDates` | Emitted when range dates are selected                  | `CustomEvent<ISelectRangeDates>` |


## Shadow Parts

| Part            | Description                      |
| --------------- | -------------------------------- |
| `"end-input"`   | The range end input container.   |
| `"start-input"` | The range start input container. |


## Dependencies

### Used by

 - [kv-calendar-advanced-date-selector](../calendar-advanced-date-selector)

### Depends on

- [kv-dropdown-base](../dropdown-base)
- [kv-text-field](../text-field)
- [kv-icon](../icon)
- [kv-calendar-range-dates-selector](../calendar-range-dates-selector)

### Graph
```mermaid
graph TD;
  kv-range-dates-select-dropdown --> kv-dropdown-base
  kv-range-dates-select-dropdown --> kv-text-field
  kv-range-dates-select-dropdown --> kv-icon
  kv-range-dates-select-dropdown --> kv-calendar-range-dates-selector
  kv-text-field --> kv-tooltip
  kv-text-field --> kv-form-label
  kv-text-field --> kv-icon
  kv-text-field --> kv-form-help-text
  kv-form-help-text --> kv-icon
  kv-calendar-range-dates-selector --> kv-calendar
  kv-calendar --> kv-action-button-icon
  kv-calendar --> kv-calendar-day
  kv-action-button-icon --> kv-action-button
  kv-action-button-icon --> kv-icon
  kv-action-button-icon --> kv-badge
  kv-calendar-advanced-date-selector --> kv-range-dates-select-dropdown
  style kv-range-dates-select-dropdown fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------


