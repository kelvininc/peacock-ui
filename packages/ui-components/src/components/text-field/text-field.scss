@import '../../assets/styles/globals';

$default-input-height: var(--input-height, $spacing-9x);
$slim-input-height: var(--input-height, $spacing-7x);

.label,
.error {
	@include kv-font-label-small-uppercase-semibold; }

.label {
	color: kv-color('text');
}

.error {
	color: kv-color('error');
	margin-right: $spacing;
}

.flex-align-center {
	display: flex;
	align-content: center;
}

.text-container {
	margin-left: $spacing-2x;

	.text-elem {
		@include kv-font-label-small-regular;
		width: 100%;
		height: 16px;
		margin-top: $spacing;

		color: kv-color('neutral-5');

		.flex-align-center {
			display: flex;
			align-content: center;
		}

		svg-icon {
			margin-right: $spacing;
		}
	}

	.has-error {
		color: kv-color('error');
	}
}

input {
	@include kv-font-span-regular;

	color: kv-color('primary');
	width: calc(100% - $spacing-4x);
	padding-left: $spacing-2x;
	padding-right: $spacing-2x;
	padding-top: 0;
	padding-bottom: 0;
	outline: 0;
	border-radius: 4px;
	border: solid 1px kv-color('neutral-2');
	height: $default-input-height;

	.has-error {
		border-color: kv-color('error');
	}

	&:disabled {
		background-color: kv-color('neutral-3');
		color: kv-color('text');
	}

	&::placeholder {
		color: kv-color('neutral-2');
	}

	&:disabled::placeholder {
		color: kv-color('neutral-5');
	}

	&:focus {
		border-color: kv-color('primary');
		color: kv-color('primary');
	}
}

.input-container {
	position: relative;
	margin-top: $spacing;

	.svg-icon {
		position: absolute;
		top: 0;
		right: $spacing-2x;
		height: 100%;
		cursor: pointer;
		user-select: none;
	}
}

.input-container-loading {

	@include ghost-animation;
	display: flex;
	height: $default-input-height;
	margin-top: $spacing;
}

.kv-slim-text-field {
	height: $slim-input-height;
}
