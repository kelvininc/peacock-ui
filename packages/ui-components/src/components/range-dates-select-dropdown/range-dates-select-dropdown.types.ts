import { ComputePositionConfig } from '@floating-ui/dom';
import { EventEmitter } from '@stencil/core';
import { ITextField } from '../../types';
import { ICalendarRangeDatesSelector, ICalendarRangeDatesSelectorEvents } from '../calendar-range-dates-selector/calendar-range-dates-selector.types';
import { IDropdownBase } from '../dropdown-base/dropdown-base.types';

export interface IRangeDatesSelectDropdown extends ICalendarRangeDatesSelector, IDropdownBase {
	/** (optional) The start date text field input custom configurations */
	startInputConfig?: Partial<ITextField>;
	/** (optional) The end date text field input custom configurations */
	endInputConfig?: Partial<ITextField>;
	/** (optional) If `true` the calendar is opened */
	isOpen?: boolean;
	/** (optional) The dropdown position options */
	dropdownPositionOptions?: Partial<ComputePositionConfig>;
	/** (optional) The text field start date format mask */
	startDateMask?: string;
	/** (optional) The text field end date format mask */
	endDateMask?: string;
	/** (optional) If `true` the dropdown will be automatically closed after start and end date is selected */
	autoClose?: boolean;
}

export interface IRangeDatesSelectDropdownEvents extends ICalendarRangeDatesSelectorEvents {
	/** Emitted when the calendar selector opens state changes */
	openStateChange: EventEmitter<boolean>;
}
