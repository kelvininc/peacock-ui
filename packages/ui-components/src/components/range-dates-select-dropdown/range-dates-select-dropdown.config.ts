import { ComputePositionConfig, flip, offset } from '@floating-ui/dom';
import { EIconName } from '../icon/icon.types';
import { EInputFieldType, EValidationState, ITextField } from '../text-field/text-field.types';

export const DEFAULT_START_DATE_INPUT_CONFIG: Partial<ITextField> = {
	placeholder: 'Select a start date',
	type: EInputFieldType.Text,
	icon: EIconName.Calendar,
	state: EValidationState.None
};

export const DEFAULT_END_DATE_INPUT_CONFIG: Partial<ITextField> = {
	placeholder: 'Select an end date',
	type: EInputFieldType.Text,
	icon: EIconName.Calendar,
	state: EValidationState.None
};

export const DEFAULT_DROPDOWN_POSITION_OPTIONS: Partial<ComputePositionConfig> = {
	placement: 'left-start',
	middleware: [
		offset(2),
		flip({
			fallbackPlacements: ['left-end', 'bottom-start', 'bottom-end', 'right-start', 'right-end', 'top-start', 'top-end']
		})
	]
};
