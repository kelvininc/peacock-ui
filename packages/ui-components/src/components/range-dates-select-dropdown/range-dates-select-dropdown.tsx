import { ComputePositionConfig } from '@floating-ui/dom';
import { Component, Element, Event, EventEmitter, h, Host, Prop } from '@stencil/core';
import { isEmpty, merge } from 'lodash-es';
import { ITextField } from '../../types';
import { IInputConfig } from '../calendar-advanced-date-selector/calendar-advanced-date-selector.types';
import { formatDateTime } from '../../utils/date.helper';
import { EIconName } from '../icon/icon.types';
import { DEFAULT_DROPDOWN_POSITION_OPTIONS, DEFAULT_END_DATE_INPUT_CONFIG, DEFAULT_START_DATE_INPUT_CONFIG } from './range-dates-select-dropdown.config';
import { IRangeDatesSelectDropdown, IRangeDatesSelectDropdownEvents } from './range-dates-select-dropdown.types';
import { ISelectRangeDates } from '../calendar-range-dates-selector/calendar-range-dates-selector.types';
import { SelectedRange } from '../../types';

@Component({
	tag: 'kv-range-dates-select-dropdown',
	styleUrl: 'range-dates-select-dropdown.scss',
	shadow: false
})
export class KvRangeDatesSelectDropdown implements IRangeDatesSelectDropdown, IRangeDatesSelectDropdownEvents {
	/** @inheritdoc */
	@Prop({ reflect: false }) startInputConfig?: IInputConfig = {};
	/** @inheritdoc */
	@Prop({ reflect: false }) endInputConfig?: IInputConfig = {};
	/** @inheritdoc */
	@Prop({ reflect: true, mutable: true }) isOpen?: boolean = false;
	/** @inheritdoc */
	@Prop({ reflect: true }) selectedRangeDates?: SelectedRange;
	/** @inheritdoc */
	@Prop({ reflect: false }) initialDate?: string;
	/** @inheritdoc */
	@Prop({ reflect: false }) disabledDates?: string[] = [];
	/** @inheritdoc */
	@Prop({ reflect: false }) minDate?: string;
	/** @inheritdoc */
	@Prop({ reflect: false }) maxDate?: string;
	/** @inheritdoc */
	@Prop({ reflect: false }) dropdownPositionOptions?: Partial<ComputePositionConfig> = DEFAULT_DROPDOWN_POSITION_OPTIONS;
	/** @inheritdoc */
	@Prop({ reflect: false }) autoClose?: boolean = true;
	/** @inheritdoc */
	@Prop({ reflect: true }) disabled?: boolean = false;

	/** @inheritdoc */
	@Event({ bubbles: false }) openStateChange: EventEmitter<boolean>;
	/** @inheritdoc */
	@Event() selectRangeDates: EventEmitter<ISelectRangeDates>;

	@Element() el: HTMLKvRangeDatesSelectDropdownElement;

	private onToggleOpenState = () => {
		if (!this.disabled) {
			this.isOpen = !this.isOpen;
			this.openStateChange.emit(this.isOpen);
		}
	};

	private getSelectedRangeRanges = (): SelectedRange => {
		return this.selectedRangeDates ?? [];
	};

	private getStartInputConfig = (): Partial<ITextField> => {
		return merge({}, DEFAULT_START_DATE_INPUT_CONFIG, { disabled: this.disabled }, this.startInputConfig);
	};

	private getEndInputConfig = (): Partial<ITextField> => {
		return merge({}, DEFAULT_END_DATE_INPUT_CONFIG, { disabled: this.disabled }, this.endInputConfig);
	};

	public getFormattedSelectedStartDate = (): string | undefined => {
		const [startDate] = this.getSelectedRangeRanges();
		if (!isEmpty(startDate)) {
			return formatDateTime(startDate, this.startInputConfig.dateMask);
		}
	};

	public getFormattedSelectedEndDate = (): string | undefined => {
		const [, endDate] = this.getSelectedRangeRanges();

		if (!isEmpty(endDate)) {
			return formatDateTime(endDate, this.endInputConfig.dateMask);
		}
	};

	public isStartSingleDateSelectDropdownFocus = (): boolean => {
		const [startDate] = this.getSelectedRangeRanges();

		return this.isOpen && isEmpty(startDate);
	};

	public isEndSingleDateSelectDropdownFocus = (): boolean => {
		const [startDate, endDate] = this.getSelectedRangeRanges();

		return this.isOpen && !isEmpty(startDate) && isEmpty(endDate);
	};

	private onOpenStateChange = ({ detail: openState }: CustomEvent<boolean>) => {
		this.isOpen = openState;
	};

	private onSelectRangeDates = ({ detail }: CustomEvent<ISelectRangeDates>) => {
		this.selectRangeDates.emit(detail);
		const { payload: newSelectRangeDates } = detail;
		const [startDate, endDate] = newSelectRangeDates;

		if (!this.autoClose) {
			return;
		}

		if (this.isOpen && !isEmpty(startDate) && !isEmpty(endDate)) {
			this.onToggleOpenState();
		}
	};

	render() {
		return (
			<Host>
				<div class="range-dates-select-dropdown">
					<kv-dropdown-base isOpen={this.isOpen} options={this.dropdownPositionOptions} onOpenStateChange={this.onOpenStateChange}>
						<div slot="action" class="inputs-container">
							<div class="start-single-date-select-dropdown">
								<kv-text-field
									{...this.getStartInputConfig()}
									value={this.getFormattedSelectedStartDate()}
									onFieldClick={this.onToggleOpenState}
									readonly
									actionIcon={this.isStartSingleDateSelectDropdownFocus() ? EIconName.ArrowDropUp : EIconName.ArrowDropDown}
									forcedFocus={this.isStartSingleDateSelectDropdownFocus()}
								/>
							</div>
							<div class="end-single-date-select-dropdown">
								<kv-text-field
									{...this.getEndInputConfig()}
									value={this.getFormattedSelectedEndDate()}
									onFieldClick={this.onToggleOpenState}
									readonly
									actionIcon={this.isEndSingleDateSelectDropdownFocus() ? EIconName.ArrowDropUp : EIconName.ArrowDropDown}
									forcedFocus={this.isEndSingleDateSelectDropdownFocus()}
								/>
							</div>
						</div>
						<div slot="list" class="range-dates-select-dropdown-calendar-container">
							<kv-calendar-range-dates-selector
								selectedRangeDates={this.getSelectedRangeRanges()}
								initialDate={this.initialDate}
								disabledDates={this.disabledDates}
								minDate={this.minDate}
								maxDate={this.maxDate}
								onSelectRangeDates={this.onSelectRangeDates}
							/>
						</div>
					</kv-dropdown-base>
				</div>
			</Host>
		);
	}
}
