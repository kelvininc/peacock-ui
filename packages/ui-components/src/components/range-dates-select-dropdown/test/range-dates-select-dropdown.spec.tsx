import { SpecPage, h } from '@stencil/core/internal';

import { KvCalendarRangeDatesSelector } from '../../calendar-range-dates-selector/calendar-range-dates-selector';
import { KvRangeDatesSelectDropdown } from '../range-dates-select-dropdown';
import { KvTextField } from '../../text-field/text-field';
import { newSpecPage } from '@stencil/core/testing';

describe('Range Dates Input (unit tests)', () => {
	let page: SpecPage;
	let component: KvRangeDatesSelectDropdown;

	describe('when uses default props', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvRangeDatesSelectDropdown, KvTextField, KvCalendarRangeDatesSelector],
				html: '<kv-range-dates-select-dropdown></kv-range-dates-select-dropdown>'
			});
			component = page.rootInstance;
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});

		it('should initialize `isOpen` with false', () => {
			expect(component.isOpen).toBe(false);
		});
	});

	describe('when `getFormattedSelectedStartDate` is called', () => {
		describe('and there is no selected range dates', () => {
			let result: string | undefined;

			beforeEach(async () => {
				page = await newSpecPage({
					components: [KvRangeDatesSelectDropdown, KvTextField, KvCalendarRangeDatesSelector],
					html: '<kv-range-dates-select-dropdown></kv-range-dates-select-dropdown>'
				});
				component = page.rootInstance;
				result = component.getFormattedSelectedStartDate();
			});

			it('should return undefined', () => {
				expect(result).toBeUndefined();
			});
		});

		describe('and there is a range start selected date', () => {
			describe('and a custom date mask is defined', () => {
				let result: string | undefined;

				beforeEach(async () => {
					page = await newSpecPage({
						components: [KvRangeDatesSelectDropdown, KvTextField, KvCalendarRangeDatesSelector],
						template: () => (
							<kv-range-dates-select-dropdown
								selectedRangeDates={['2020-04-02', '2020-04-05']}
								startInputConfig={{ dateMask: 'Do MMM, YYYY' }}
							></kv-range-dates-select-dropdown>
						)
					});
					component = page.rootInstance;
					result = component.getFormattedSelectedStartDate();
				});

				it('should return start date formatted', () => {
					expect(result).toEqual('2nd Apr, 2020');
				});
			});

			describe('and a custom date mask is not defined', () => {
				let result: string | undefined;

				beforeEach(async () => {
					page = await newSpecPage({
						components: [KvRangeDatesSelectDropdown, KvTextField, KvCalendarRangeDatesSelector],
						template: () => <kv-range-dates-select-dropdown selectedRangeDates={['2020-04-02', '2020-04-05']}></kv-range-dates-select-dropdown>
					});
					component = page.rootInstance;
					result = component.getFormattedSelectedStartDate();
				});

				it('should return start date formatted', () => {
					expect(result).toEqual('2020-04-02 00:00:00');
				});
			});
		});
	});

	describe('when `getFormattedSelectedEndDate` is called', () => {
		describe('and there is no selected range dates', () => {
			let result: string | undefined;

			beforeEach(async () => {
				page = await newSpecPage({
					components: [KvRangeDatesSelectDropdown, KvTextField, KvCalendarRangeDatesSelector],
					html: '<kv-range-dates-select-dropdown></kv-range-dates-select-dropdown>'
				});
				component = page.rootInstance;
				result = component.getFormattedSelectedEndDate();
			});

			it('should return undefined', () => {
				expect(result).toBeUndefined();
			});
		});

		describe('and there is a range end selected date', () => {
			describe('and a custom date mask is defined', () => {
				let result: string | undefined;

				beforeEach(async () => {
					page = await newSpecPage({
						components: [KvRangeDatesSelectDropdown, KvTextField, KvCalendarRangeDatesSelector],
						template: () => (
							<kv-range-dates-select-dropdown
								selectedRangeDates={['2020-04-02', '2020-04-05']}
								endInputConfig={{ dateMask: 'Do MMM, YYYY' }}
							></kv-range-dates-select-dropdown>
						)
					});
					component = page.rootInstance;
					result = component.getFormattedSelectedEndDate();
				});

				it('should return end date formatted', () => {
					expect(result).toEqual('5th Apr, 2020');
				});
			});
			describe('and a custom date mask is not defined', () => {
				let result: string | undefined;

				beforeEach(async () => {
					page = await newSpecPage({
						components: [KvRangeDatesSelectDropdown, KvTextField, KvCalendarRangeDatesSelector],
						template: () => <kv-range-dates-select-dropdown selectedRangeDates={['2020-04-02', '2020-04-05']}></kv-range-dates-select-dropdown>
					});
					component = page.rootInstance;
					result = component.getFormattedSelectedEndDate();
				});

				it('should return end date formatted', () => {
					expect(result).toEqual('2020-04-05 00:00:00');
				});
			});
		});
	});

	describe('when `isStartSingleDateSelectDropdownFocus` is called', () => {
		describe('and the calendar is closed', () => {
			let result: boolean;

			beforeEach(async () => {
				page = await newSpecPage({
					components: [KvRangeDatesSelectDropdown, KvTextField, KvCalendarRangeDatesSelector],
					template: () => <kv-range-dates-select-dropdown isOpen={false}></kv-range-dates-select-dropdown>
				});
				component = page.rootInstance;
				result = component.isStartSingleDateSelectDropdownFocus();
			});

			it('should return false', () => {
				expect(result).toBe(false);
			});
		});
	});

	describe('when `isEndSingleDateSelectDropdownFocus` is called', () => {
		describe('and the calendar is closed', () => {
			let result: boolean;

			beforeEach(async () => {
				page = await newSpecPage({
					components: [KvRangeDatesSelectDropdown, KvTextField, KvCalendarRangeDatesSelector],
					template: () => <kv-range-dates-select-dropdown isOpen={false}></kv-range-dates-select-dropdown>
				});
				component = page.rootInstance;
				result = component.isEndSingleDateSelectDropdownFocus();
			});

			it('should return false', () => {
				expect(result).toBe(false);
			});
		});
	});
});
