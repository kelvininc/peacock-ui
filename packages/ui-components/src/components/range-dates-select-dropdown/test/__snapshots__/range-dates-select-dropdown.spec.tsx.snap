// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Range Dates Input (unit tests) when uses default props should match the snapshot 1`] = `
<kv-range-dates-select-dropdown>
  <div class="range-dates-select-dropdown">
    <kv-dropdown-base>
      <div class="inputs-container" slot="action">
        <div class="start-single-date-select-dropdown">
          <kv-text-field action-icon="kv-arrow-drop-down" icon="kv-calendar" input-mask-regex="" placeholder="Select a start date" readonly="" state="none" type="text" value="">
            <mock:shadow-root>
              <kv-tooltip position="top-start" truncate="">
                <div class="text-field-container">
                  <kv-form-label></kv-form-label>
                  <div class="input-container input-container--size-large">
                    <input class="left-slotted right-slotted" placeholder="Select a start date" readonly="" type="text" value="">
                    <div class="left-slot-container">
                      <slot name="left-slot">
                        <kv-icon exportparts="icon" name="kv-calendar"></kv-icon>
                      </slot>
                    </div>
                    <div class="right-icon-container">
                      <kv-icon name="kv-arrow-drop-down"></kv-icon>
                    </div>
                  </div>
                  <kv-form-help-text state="none"></kv-form-help-text>
                </div>
              </kv-tooltip>
            </mock:shadow-root>
          </kv-text-field>
        </div>
        <div class="end-single-date-select-dropdown">
          <kv-text-field action-icon="kv-arrow-drop-down" icon="kv-calendar" input-mask-regex="" placeholder="Select an end date" readonly="" state="none" type="text" value="">
            <mock:shadow-root>
              <kv-tooltip position="top-start" truncate="">
                <div class="text-field-container">
                  <kv-form-label></kv-form-label>
                  <div class="input-container input-container--size-large">
                    <input class="left-slotted right-slotted" placeholder="Select an end date" readonly="" type="text" value="">
                    <div class="left-slot-container">
                      <slot name="left-slot">
                        <kv-icon exportparts="icon" name="kv-calendar"></kv-icon>
                      </slot>
                    </div>
                    <div class="right-icon-container">
                      <kv-icon name="kv-arrow-drop-down"></kv-icon>
                    </div>
                  </div>
                  <kv-form-help-text state="none"></kv-form-help-text>
                </div>
              </kv-tooltip>
            </mock:shadow-root>
          </kv-text-field>
        </div>
      </div>
      <div class="range-dates-select-dropdown-calendar-container" slot="list">
        <kv-calendar-range-dates-selector>
          <mock:shadow-root>
            <kv-calendar exportparts="calendar-container,month-container,day-container"></kv-calendar>
          </mock:shadow-root>
        </kv-calendar-range-dates-selector>
      </div>
    </kv-dropdown-base>
  </div>
</kv-range-dates-select-dropdown>
`;
